### PR TITLE
fix: exit HTML fullscreen when window leaves fullscreen

### PR DESCRIPTION
**Pull request by miniak**
Fixes https://github.com/electron/electron/issues/12653

### CODE REVIEW

**Comment by deepak1556** [comment]
@miniak content layer checks for the fullscreen state of the page with `WebContentsDelegate::IsFullscreenForTabOrPending` and we have been returning the wrong state https://github.com/electron/electron/blob/master/atom/browser/common_web_contents_delegate.cc#L273, also this bug is present since the original implementation. The fix here would be to return the fullscreen state of the owner window using `owner_window_->IsFullscreen()`

**Comment by zcbenz** [comment]
> @zcbenz any idea how to do this in the C++ code directly?

You can add an override in `atom_api_browser_window.cc`:

```c++
BrowserWindow::OnWindowLeaveFullScreen() {
  TopLevelWindow::OnWindowLeaveFullScreen();
  web_contents()->...
}
```

But WebContents does not have an API to execute script directly, and `webContents.executeJavaScript` is implemented in JavaScript, so you have to send IPC messages to implement the same behavior.

File: lib/browser/api/browser-window.js [hunk 1]
```diff
@@ -103,6 +104,15 @@ BrowserWindow.prototype._init = function () {
     app.emit('browser-window-focus', event, this)
   })
 
+  // Exit HTML fullscreen when window leaves fullscreen
+  this.on('leave-full-screen', (event) => {
+    for (const webContents of getAllWebContents()) {
+      if (webContents.getOwnerBrowserWindow() === this) {
```
Comment (zcbenz) [inline lib/browser/api/browser-window.js]: We can use `this.webContents` here instead of searching for all WebContents.

Comment (miniak) [reply lib/browser/api/browser-window.js]: @zcbenz I know, but there might be nested webContents

**Comment by miniak** [comment]
@zcbenz by implementing in C++ I mean more like doing it without `executeJavaScript` :) I should probably look into Chromium code how they handle it in the browser code part.

**Comment by deepak1556** [comment]
> being in window fullscreen mode does not imply HTML fullscreen, however HTML fullscreen implies window fullscreen.

Yes and the value of `html_fullscreen_` is not updated correctly when the window exits fullscreen mode while in html fullscreen, which causes the content layer to read the wrong value and not update the UI.

File: lib/browser/api/browser-window.js [hunk 2]
```diff
@@ -103,6 +104,15 @@ BrowserWindow.prototype._init = function () {
...
+        webContents.executeJavaScript('document.webkitExitFullscreen()')
```
Comment (zcbenz) [inline lib/browser/api/browser-window.js]: We can check `is_html_fullscreen()` before executing the script, the API needs to be exposed in JS though.

**Pull request merged**